daemon/input.c

```diff
@@ -0,0 +1,303 @@
+	uint16_t bus;
+	uint16_t vid;
+	uint16_t pid;
+	uint16_t ver;
+	char *event = NULL;
+	char *event_path = NULL;
+
+	fp = fopen("/proc/bus/input/devices", "r");
+	IF_NULL_RETVAL(fp, NULL);
+
+	while ((n = getline(&line, &len, fp)) != -1) {
+		if (strncmp(IFACE_PREFIX, line, strlen(IFACE_PREFIX)) == 0) {
+			sscanf(line, "I: Bus=%04hx Vendor=%04hx Product=%04hx Version=%04hx\n",
+			       &bus, &vid, &pid, &ver);
+		}
+
+		if ((vid == vendor_id) && (pid == product_id) &&
+		    (strncmp(HANDLER_PREFIX, line, strlen(IFACE_PREFIX)) == 0) &&
+		    strstr(line, SYSRQ_EVENT)) {
+			event = strtok(strstr(line, EVENT), " ");
+			event_path = mem_printf("/dev/input/%s", event);
+			break;
+		}
+	}
+
+	fclose(fp);
+	mem_free(line);
+	return event_path;
+}
+
+static void
+input_cb_request_pin_start_container(int fd, unsigned events, event_io_t *io, void *data)
+{
+	static char *key = NULL;
+	static int index = 0;
+	char *tmp_key = NULL;
+
+	ASSERT(data);
+	if (events & EVENT_IO_EXCEPT) {
+		DEBUG("Event IO exception on USB pin reader callback entry");
+		goto exit;
+	} else if (events & EVENT_IO_READ) {
+		container_start_cb_data_t *cb_data = data;
+		if (!cb_data->control || !cb_data->container || !cb_data->start_params) {
+			WARN("Invalid parameters for container start callback");
+			goto exit;
+		}
+
+		struct input_event keyboard_event;
+		TRACE("Reading pin from pin reader");
+		while (read(fd, &keyboard_event, sizeof(keyboard_event)) > 0) {
+			if ((keyboard_event.type != 0x1) || (keyboard_event.value != 0x1)) {
+				continue;
+			} else if (keyboard_event.code == KEY_BACKSPACE) {
+				// Backspace was pressed, remove last entered digit
+				if (index > 0) {
+					index--;
+				}
+			} else if ((keyboard_event.code == KEY_ENTER) ||
+				   (keyboard_event.code == KEY_KPENTER)) {
+				// Enter was pressed, finish
+				if (!key) {
+					key = (char *)mem_alloc(sizeof(char));
+					*key = '\0';
+				}
+				index = 0;
+				DEBUG("Starting container with smartcard from callback");
+				if (cmld_container_start_with_smartcard(
+					    cb_data->control, cb_data->container, key) != 0) {
+					ERROR("Failed to start container %s",
+					      container_get_name(cb_data->container));
+				}
+				goto exit;
+			} else {
+				// Another key was pressed..
+				char c;
+				// ..Check if pressed key is a digit and store it
+				if (input_event_code_to_ascii_num(keyboard_event.code, &c) == 0) {
+					index++;
+					tmp_key = (char *)mem_realloc(key, index * sizeof(char));
+					if (tmp_key) {
+						key = tmp_key;
+						key[index - 1] = c;
+					} else {
+						WARN("Failed to allocate memory for pin");
+						free(key);
+						index = 0;
+						goto exit;
+					}
+				}
+			}
+		}
+	}
+
+	// Pin entry is not yet finished, just return to wait for more keystrokes
+	return;
+
+exit:
+	if (ioctl(fd, EVIOCGRAB, NULL) != 0) {
+		WARN("Failed to release exclusive access to pin reader");
+	}
+	event_remove_io(io);
+	event_io_free(io);
+	close(fd);
+	mem_free(data);
```

Comment:
mem_free(cb_data); (is the same pointer but looks nicer)

Reply:
Done